src/main/java/com/conveyal/analysis/controllers/RegionalAnalysisController.java

```diff
@@ -228,89 +402,36 @@ private Object getRegionalResults (Request req, Response res) throws IOException
-                multiCutoffKey = String.format("%s_P%d.access", regionalAnalysisId, percentile);
-                multiCutoffFileStorageKey = new FileStorageKey(RESULTS, multiCutoffKey);
-                if (fileStorage.exists(multiCutoffFileStorageKey)) {
-                    checkArgument(analysis.destinationPointSetIds.length == 1);
-                } else {
-                    // Fall back on oldest form of results, single-percentile, single-destination-grid.
-                    multiCutoffKey = regionalAnalysisId + ".access";
-                    multiCutoffFileStorageKey = new FileStorageKey(RESULTS, multiCutoffKey);
-                    if (fileStorage.exists(multiCutoffFileStorageKey)) {
-                        checkArgument(analysis.travelTimePercentiles.length == 1);
-                        checkArgument(analysis.destinationPointSetIds.length == 1);
-                    } else {
-                        throw AnalysisServerException.notFound("Cannot find original source regional analysis output.");
-                    }
-                }
-            }
-            LOG.debug("Single-cutoff grid {} not found on S3, deriving it from {}.", singleCutoffKey, multiCutoffKey);
-
-            InputStream multiCutoffInputStream = new FileInputStream(fileStorage.getFile(multiCutoffFileStorageKey));
-            Grid grid = new SelectingGridReducer(cutoffIndex).compute(multiCutoffInputStream);
-
-            File localFile = FileUtils.createScratchFile(format.toString());
-            FileOutputStream fos = new FileOutputStream(localFile);
-
-            switch (format) {
-                case GRID:
-                    grid.write(new GZIPOutputStream(fos));
-                    break;
-                case PNG:
-                    grid.writePng(fos);
-                    break;
-                case GEOTIFF:
-                    grid.writeGeotiff(fos);
-                    break;
-            }
-
-            fileStorage.moveIntoStorage(singleCutoffFileStorageKey, localFile);
-        }
-        return JsonUtil.toJsonString(
-                JsonUtil.objectNode().put("url", fileStorage.getURL(singleCutoffFileStorageKey))
+        // Process has a lot of overhead: UI contacts backend, backend calls S3, backend responds to UI, UI contacts S3.
+        FileStorageKey singleCutoffFileStorageKey = getSingleCutoffGrid(
+                analysis, cutoffIndex, destinationPointSetId, percentile, format
         );
+        res.type(APPLICATION_JSON.asString());
+        // Substitute human readable name and shorten destination UUID.
+        // TODO better system for resolving UUID to human readable names in single and multi-grid cases
+        //      Or maybe only allow multi-grid download for saving by end user, and single grid is purely internal.
+        int firstUnderscore = singleCutoffFileStorageKey.path.indexOf('_');
+        int secondUnderscore = singleCutoffFileStorageKey.path.indexOf('_', firstUnderscore + 1);
```

Comment:
This parses out the destination id, cutoff, and percentile from `singleCutoffFileStorageKey.path`. 

Is there a reason for not using `cutoffIndex`, `destinationPointSetId`, `percentile` directly? Perhaps to avoid replicating the logic for the annotations (`C`, `P`)?